File: Programmers_java/순위.java
class Solution {
    public int solution(int n, int[][] results) {
        final int INF = 987654321;
        int[][] players = new int[n + 1][n + 1];
        int answer = 0;

        for (int i = 1; i <= n; i++) {  // 초기 세팅
            for (int j = 1; j <= n; j++) {
                if (i == j) {
                    players[i][j] = 0;
                } else {
                    players[i][j] = INF;
                }
            }
        }

        for (int[] result : results) {  // 경기 결과 대입
            int winner = result[0];
            int loser = result[1];
            players[winner][loser] = 1;
        }

        for (int k = 1; k <= n; k++) {  // 플로이드 워셜
            for (int i = 1; i <= n; i++) {
                for (int j = 1; j <= n; j++) {
                    players[i][j] = Math.min(players[i][j], players[i][k] + players[k][j]);
                }
            }
        }

        boolean[] flag = new boolean[n + 1];
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= n; j++) {
                if (players[i][j] == INF) {   // 두 선수 사이에 승패 결과가 없으면
                    if (players[j][i] == INF) {
                        flag[i] = true;
                        break;
                    }
                }
            }
        }

        for (int i = 1; i <= n; i++) {  // 순위를 알 수 있는 선수
            if (!flag[i]) {
                answer++;
            }
        }

        return answer;
    }
}

File: Programmers_python/순위.py
INF = int(1e9)

def solution(n, results):
    answer = 0
    game = [[INF] * (n+1) for _ in range(n+1)]
    ranking = [True] * (n+1)

    for i in range(1, n+1):
        for j in range(1, n+1):
            if i==j:
                game[i][j] = 0
                
    for winner, loser in results:
         game[winner][loser] = 1
                
    for k in range(1, n+1):
        for i in range(1, n+1):
            for j in range(1, n+1):
                game[i][j] = min(game[i][j], game[i][k]+game[k][j])
    
    for i in range(1, n+1):
        able = True
        for j in range(1, n+1):
            if game[i][j] == INF and game[j][i] == INF:
                able = False
                break
        ranking[i] = able
            
    return ranking[1:].count(True)
